py/ringbuf: Add micropython.ringbuffer() interface for general use.

andrewleech opened the pull request:
I've had a number of desirable use cases for a python ringbuffer over the years, usually relating to logging and/or connecting streams together in various ways. 

This PR adds an optional python interface to the existing C ringbuffer library.

---
This work was funded by Planet Innovation.

pi-anl committed:
py/objringio: Add micropython.RingIO() interface for general use.

This commit adds a new `RingIO` type which exposes the internal ring-buffer
code for general use in Python programs.  It has the stream interface
making it similar to `StringIO` and `BytesIO`, except `RingIO` has a fixed
buffer size and is automatically safe when reads and writes are in
different threads or an IRQ.

This new type is enabled at the "extra features" ROM level.

Signed-off-by: Andrew Leech <andrew.leech@planetinnovation.com.au>

robert-hh commented:
Looks interesting. What is the advantage over StringIO?

andrewleech commented:
Ummm, I feel a bit like you've exposed a lack of mental agility on my part here @robert-hh I feel like I've never quite equated the two being basically there same!

In an embarrassingly similar vein I raised this just today https://github.com/micropython/micropython/discussions/9456 but never thought of using this new ringbuffer API on an existing buffer as being the same thing... I even had an earlier version of this PR where a bytearray could be passed in for reuse but couldn't quite think of a good reason to use it. Maybe I'm multitasking too much lately.

Longer term, I've previously had offline discussions about a semi-persistent logging buffer in ram that isn't explicitly wiped at startup (would require linker script updates to do this) such that logging buffered to it could be retrieved after a watchdog reset etc. I was thinking this API may be needed there too... I think the entire ringbuffer object would need to be statically placed in this fixed location in C to also maintain the head/tail idx variables.

Back to your question though, I don't think StringIO/BytesIO have a mechanism to provide a fixed length buffer. if you try to write more than they currently have space for it'll automatically re-allocate a larger buffer and copy the contents. This can be undesirable when buffering larger amounts of data on smaller parts, or when speed is needed, or buffering out of interrupts etc.

It doesn't look like cpython has a way to use BytesIO with fixed length either.

peterhinch commented:
This could be very useful. Key properties of a ringbuf are:
 * Inherent thread safety.
 * Guaranteed non-allocating writes.
 
These allow writing in a hard ISR and reading in the main code (and _vice versa_).

stephanelsmith commented:
> This could be very useful. Key properties of a ringbuf are:
>  * Inherent thread safety.
>  * Guaranteed non-allocating writes.
>  
> These allow writing in a hard ISR and reading in the main code (and _vice versa_).

Love this idea

glenn20 commented:
> Back to your question though, I don't think StringIO/BytesIO have a mechanism to provide a fixed length buffer. if you try to write more than they currently have space for it'll automatically re-allocate a larger buffer and copy the contents. This can be undesirable when buffering larger amounts of data on smaller parts, or when speed is needed, or buffering out of interrupts etc.

Besides the fixed length issue, as I understand it, StringIO/BytesIO don't provide support for the desired ring buffer mode of operation.

andrewleech commented:
> StringIO/BytesIO don't provide support for the desired ring buffer mode of operation

Yes of course. The other day I was thinking this PR has a limitation in that a write cannot overwrite unread data, which is sometimes desirable (only buffer latest X bytes of incoming data).

The real benefit of ringbuffer is that the writer has a different `tell()` to the reader so it's arguably more like a socket pair than a single BytesIO which read/write at the same location.

peterhinch commented:
If a write can overwrite unread data I think this would compromise thread-safety.

glenn20 commented:
> If a write can overwrite unread data I think this would compromise thread-safety.

The underlying ringbuf_t structure is inherently threadsafe for the specific case where a single thread (or ISR) is always writing and another is reading. It is used for this purpose in the bluetooth module. 

This is very useful for my particular use case: writing data in C space from an ISR (or high priority wifi task) and consuming it in python user space.

It is not threadsafe in the general case.

peterhinch commented:
@glenn20 I should have clarified that this is the sense in which I meant it. The ISR use case you cite is widespread and important.

If, as @andrewleech proposed, a mode is supported where the write pointer can overwrite unread data, in that mode the limited thread safety is lost.

andrewleech commented:
> If a mode is supported where the write pointer can overwrite unread data, in that mode the limited thread safety is lost.

To be clear the underlying C ringbuffer implementation doesn't support this overwrite usage, nor do I have any plans to add it. It's just something that some people think of with generic ringbuffers.

glenn20 commented:
@andrewleech  : FYI - I have some work in progress which add a few extra features which are helpful for my use case, including:

- Add `ringbuf_read(r, data, size)` and `ringbuf_write(r, data, size)` which copy data to/from the ringbuffer with memcpy(), rather than byte or word at a time (I want to maximise throughput).
- Add `ringbuf_read_wait(r, data, size, timeout_ms)` and `ringbuf_write_wait(r, data, size, timeout_ms)` which provide C api to blocking IO with timeout.
- Add `mp_obj_new_ringbuffer(buf,  size, timeout_ms)` to provide C api to creating new ringbuffer objects.
- Add buffer protocol support for direct access to the underlying buffer.

You can see the changes at: https://github.com/glenn20/micropython/commit/158df921001cfe0746534efd5df21d2b1e946b72 and https://github.com/glenn20/micropython/commit/10fc526c2de4b42897819071e8754b8b5d69ec8a

If you have an interest in folding any of these into your PR, let me know.

I have this all working in a dev version of the espnow module, where the ringbuffer object is created within my module and exposed to the users as an attribute. The ringbuffer is populated with incoming data from a high priority wifi task on the esp32 and readout from the buffer can now be entirely handled in python code - allowing me to significantly reduce the C footprint of the module and support flexible user-controlled readout strategies. 

Thanks for your very timely publication of this PR ;).

andrewleech commented:
Thanks @glenn20 I had been toying with the idea of adding functions to read/write in chunks rather than byte-at-a-time but figured if I did that I'd need to do some benchmarks to check it really was a lot faster and didn't feel like doing the work :-)

I had wondered about making it buffer api compatible but thought it was less important if a buffer was passed in as it'd already be available for access - but adding this interface is important for one created in C, or convenient for one created by length.

That being said I'm still unsure about the utility of directly accessing the buffer in many cases, as the data in generally wouldn't make sense once it had written around then end of the ring. 
If the ringuffer is instead being used as a stream read/write once to a buffer ie https://github.com/orgs/micropython/discussions/9456 then this makes sense. For this usage I've added a `reset()` function to the python class to zero out both iput/iget.

I was also thinking I'd like to make the ringbuffer code compatible with the [power-of-2 variant of rungbuffers](https://github.com/micropython/micropython/pull/4617#issuecomment-475026061) where you don't lose a byte of buffer to tracking. Was thinking it could auto-select the algorith used depending on whether the supplied buffer meets this criteria. This would make extra accessor functions more complicated though as they'd also have to support both algorithms, so again I'll need to think if the benefits are worth the (code/flash) costs.

It's also worth thinking/talking about whether it'd be better to be blocking or not by default; timeout==0 ot timeout==MAX

ma261065 commented:
> The ringbuffer is populated with incoming data from a high priority wifi task on the esp32 and readout from the buffer
> can now be entirely handled in python code - allowing me to significantly reduce the C footprint of the module 
> and support flexible user-controlled readout strategies.

Do you happen to have a link to some code where you set up a high-priority wifi task to populate a buffer? This is something that I need to do in a project of mine, so it would be very useful not to have to reinvent the wheel. Much appreciated if you could share...

glenn20 commented:
> > The ringbuffer is populated with incoming data from a high priority wifi task on the esp32 and readout from the buffer
> > can now be entirely handled in python code - allowing me to significantly reduce the C footprint of the module
> > and support flexible user-controlled readout strategies.
> 
> Do you happen to have a link to some code where you set up a high-priority wifi task to populate a buffer? This is something that I need to do in a project of mine, so it would be very useful not to have to reinvent the wheel. Much appreciated if you could share...

Sorry - I didn't setup the high priority task. On the ESP32, the wifi handling code is managed with a set of high-priority threads. Espressif's ESP-NOW api includes a callback when a message is received. The callback I provide is called from one of the wifi high-priority tasks. It is a pradigm that is very similar to running the callback from an ISR.

projectgus commented:
This is an automated heads-up that we've just merged a Pull Request
that removes the STATIC macro from MicroPython's C API.

See https://github.com/micropython/micropython/pull/13763

A search suggests this PR might apply the STATIC macro to some C code. If it
does, then next time you rebase the PR (or merge from master) then you should
please replace all the `STATIC` keywords with `static`.

Although this is an automated message, feel free to @-reply to me directly if
you have any questions about this.

andrewleech commented:
I've rebased this with some rework to:
* defer to the (somewhat newer) `ringbuf_put_bytes()` / `ringbuf_gett_bytes()` functions.  
* remove the timeout functions, these were adding bulk without having a real use case (select / asyncio is better for this)

This has made this a notably smaller change than it used to be.

I've also reaslised a pair of these can be used similar to a socket pair, or a mock/virtual uart: https://github.com/micropython/micropython-lib/pull/907

andrewleech commented on py/ringbuf.c:
```diff
@@ -118,3 +118,114 @@ int ringbuf_put_bytes(ringbuf_t *r, const uint8_t *data, size_t data_len) {
+    uint8_t *dest = buf_in;
+    for (mp_uint_t i = 0; i < size; i++) {
+        *dest++ = ringbuf_get(&(self->ringbuffer));
+    }
```
@dpgeorge I realised a good chunk of the extra size in this PR was due to me "reusing" the `ringbuf_get_bytes()` and `ringbuf_put_bytes()` [functions above](https://github.com/micropython/micropython/blob/a20fca8c04f95fbf9420d3c6926fd59aa6086c83/py/ringbuf.c#L82)

I hadn't realised these aren't currently used anywhere else other than in espnow, so are stripped out of most builds currently.

Replacing their use with a simple byte by byte loop is probably slightly slower, but considerably smaller.

AmirHmZz replied:
@andrewleech Ring buffer has many use cases in our projects in the core components. Making these simple functions slower will effect not only esp now but other use cases too.
Beside that, In many chips increasing firmware size doesn't matter that much. Maybe we can choose between the slow and fast version with a flag?

andrewleech replied:
I haven't touched the function used by espnow.

I might need to think about how to profile the performance, because I'm not confident the current byte by byte method would be much slower than the memcpy one earlier, especially with the extra "stuff" going on around the memcpy version above. 

Currently all stdio goes via the same byte by byte ringbuffer approach and it doesn't seem too bad ...

andrewleech commented:
I've performed some performance testing to compare the simple byte-by-byte loop to the memcpy based method with the following script.
``` python
import micropython

try:
    micropython.ringbuffer
except AttributeError:
    print("SKIP")
    raise SystemExit


def speed(rb_size, bytes_tx):
    try:
        from time import ticks_us, ticks_diff
    except ImportError:
        from time import perf_counter

        ticks_us = lambda: int(perf_counter() * 1000000)
        ticks_diff = lambda a, b: a - b
    
    ba = bytearray(rb_size)
    mv = memoryview(ba)
    rb = micropython.ringbuffer(rb_size)
    sent = 0
    step = 1
    start = ticks_us()
    while sent < bytes_tx:
        sent += rb.write(mv[:step])
        rb.read()
        step += 1
        if step >= rb_size:
            step = 1
    end = ticks_us()
    taken = ticks_diff(end, start) / 1000
    print(f"taken {taken:0.1f}ms")
 
 
speed(128, 1*1024*1024)
```

It creates a ringbuffer (128 bytes deep) and write/read 1MB of data through it. Data is transferred in different sized chunks (incrementing step size) to sort of "ransomise" the wrap-over the end of the buffer. I ran this with a few different ringbuffer sizes / amounts of data on unix, rp2 and stm ports and yes there is still a performance gain using memcpy.

On unix with 5MB data I see 67ms (b-b-b) vs 47ms (memcpy) - memcpy is 30% faster than byte by byte.
On rp2040 with 1MB I get 1439.5ms (b-b-b) vs 1092.1ms (memcpy) - memcpy is 25% faster than byte by byte.
On pybd with 1MB I get 732.3ms (b-b-b) vs 514.1ms (memcpy) - memcpy is 30% faster than byte by byte.
On stm32wb55 with 1MB I get 2989.7ms (b-b-b) vs 2505.2ms (memcpy) - memcpy is 16% faster than byte by byte.

I also tested just inlining the important parts of the memcpy based routine into the python class read/write functions and that dropped to size increase from ~160 bytes down to just 40 bytes (on stm32). So yeah, I reckon the 16% - 30% improvement in performance is probably worth a 40 byte increase.

dpgeorge commented on py/modmicropython.c:
```diff
@@ -203,6 +204,9 @@ static const mp_rom_map_elem_t mp_module_micropython_globals_table[] = {
     { MP_ROM_QSTR(MP_QSTR_schedule), MP_ROM_PTR(&mp_micropython_schedule_obj) },
     #endif
+    #if MICROPY_PY_MICROPYTHON_RINGBUFFER
+    { MP_ROM_QSTR(MP_QSTR_ringbuffer), MP_ROM_PTR(&mp_type_micropython_ringbuffer) },
```
I think we should call the type `RingIO`.

andrewleech replied:
Oh yes, that fits!

dpgeorge commented on py/mpconfig.h:
```diff
@@ -981,6 +981,11 @@ typedef double mp_float_t;
 
+// Support for micropython.ringbuffer()
+#ifndef MICROPY_PY_MICROPYTHON_RINGBUFFER
+#define MICROPY_PY_MICROPYTHON_RINGBUFFER (MICROPY_CONFIG_ROM_LEVEL_AT_LEAST_EXTRA_FEATURES)
```
Please move this to just after `MICROPY_PY_MICROPYTHON_HEAP_LOCKED`.

dpgeorge commented on py/ringbuf.c:
```diff
@@ -75,14 +75,8 @@ int ringbuf_put16(ringbuf_t *r, uint16_t v) {
-    if (ringbuf_avail(r) < data_len) {
-        return (r->size <= data_len) ? -2 : -1;
-    }
+static inline __attribute__((always_inline)) void _ringbuf_memcpy_get(ringbuf_t *r, uint8_t *data, size_t data_len) {
```
Please use `MP_ALWAYSINLINE`.

dpgeorge commented on py/ringbuf.c:
```diff
@@ -116,5 +103,153 @@ int ringbuf_put_bytes(ringbuf_t *r, const uint8_t *data, size_t data_len) {
+#include "py/mphal.h"
+
+#ifndef MICROPY_PY_MICROPYTHON_RINGBUFFER_USE_MEMCPY
+#define MICROPY_PY_MICROPYTHON_RINGBUFFER_USE_MEMCPY (1)
```
I don't think it's worth providing this option to turn memcpy optimisation off.  Just always enable it, ie always use memcpy.  It doesn't cost much extra and is faster (and you want fast when using a ringbuffer).

dpgeorge commented on py/ringbuf.c:
```diff
@@ -116,5 +103,153 @@ int ringbuf_put_bytes(ringbuf_t *r, const uint8_t *data, size_t data_len) {
+            return ret;
+        }
+        case MP_STREAM_CLOSE:
+        case MP_STREAM_FLUSH:
```
You can remove `MP_STREAM_FLUSH`.

dpgeorge commented on py/ringbuf.c:
```diff
@@ -116,5 +103,153 @@ int ringbuf_put_bytes(ringbuf_t *r, const uint8_t *data, size_t data_len) {
+}
+static MP_DEFINE_CONST_FUN_OBJ_1(micropython_ringbuffer_any_obj, micropython_ringbuffer_any);
+
+
```
Please only use a single blank line (also below).

peterhinch commented:
To add a general comment, there is a use case for a ringbuf where the data source is continuous and cannot be interrupted. The data sink may have periods where it cannot maintain the rate: if the buffer fills, data must be discarded. A suitably configured ringbuf ensures that the oldest data is lost.

An option to specify this mode might be useful.

andrewleech commented:
> A suitably configured ringbuf ensures that the oldest data is lost.
> 
> An option to specify this mode might be useful.

Yeah interesting point, currently it only supports "newest data being lost", not oldest. 

I don't think the existing C implementation supports "discard oldest", though I guess it just requires the put function to move the get pointer along as well as the put pointer when it's full? I presume this also reduces its "thread safety" aka interrupt safety... 
Could be somewhat more tricky with this memcpy variant, as the read could be half way through a large memcpy when the interrupt comes in and move all the pointers, resulting in corrupted data when the read memcpy resumes...

Maybe there's a better way to implement it? Maybe in overwrite mode byte-by-byte would be safer (as long as you can get the order right of when pointers are moved etc)

andrewleech commented:
Thanks for the review @dpgeorge ! I've addressed your comments in the latest push. 
I'm getting a fail on the codecov/patch now  (81.25% of diff hit (target 98.43%))
 due to changes not covered by tests:
```
py/ringbuf.c#L112-L114
py/ringbuf.c#L116
py/ringbuf.c#L124-L126
py/ringbuf.c#L128-L129
```
These the remainder of the existing `ringbuf_get_bytes()` / `ringbuf_put_bytes()` functions. They're part of the diff now as I pulled some of their functionality into common inline functions. These two functions are only used in the espnow module so I don't know if it's feasible to get CI test coverage on them?

dpgeorge commented on py/modmicropython.c:
```diff
@@ -200,6 +201,9 @@ static const mp_rom_map_elem_t mp_module_micropython_globals_table[] = {
     { MP_ROM_QSTR(MP_QSTR_kbd_intr), MP_ROM_PTR(&mp_micropython_kbd_intr_obj) },
     #endif
+    #if MICROPY_PY_MICROPYTHON_RINGIO
+    { MP_ROM_QSTR(MP_QSTR_RingIO), MP_ROM_PTR(&mp_type_micropython_ringio) },
```
I suggest this go at the end of the locals dict, classes come after functions (generally).

andrewleech replied:
Ah, I was going with basic alphabetical there! How about on the docs page? Should it be moved similarly?

dpgeorge replied:
It should also be moved in the docs.  You'll see there's an existing "Functions" heading, so please add a "Classes" heading at the end of the file, then RingIO docs.

dpgeorge replied:
It looks like `RingIO` is still not at the end of the locals dict here?

dpgeorge commented on py/ringbuf.c:
```diff
@@ -75,14 +75,8 @@ int ringbuf_put16(ringbuf_t *r, uint16_t v) {
-    if (ringbuf_avail(r) < data_len) {
-        return (r->size <= data_len) ? -2 : -1;
-    }
+static inline MP_ALWAYSINLINE void _ringbuf_memcpy_get(ringbuf_t *r, uint8_t *data, size_t data_len) {
```
I think it would be better to put these memcpy functions in `py/ringbuf.h` (and remove the leading underscore), so they can be public functions.

Then, the `RingIO` implementation can go in a separate file like `py/objringio.c`.

And, does this really need `MP_ALWAYSINLINE`?  If code size is unchanged without that attribute then I'd rather leave it out.

dpgeorge replied:
Then we can see how the CI coverage looks with these changes, it may improve.

andrewleech replied:
I was wary of making them public because without adding a length check in the function calling them you can get a buffer overrun failure.

Without MP_ALWAYSINLINE it was bigger on rp2. Even with both `inline MP_ALWAYSINLINE` rp2 size increases by 4 compared to literally copying the contents into the target function, so something about the inline settings on the compiler there is tricky perhaps.

dpgeorge replied:
My main motivation here is to put the implementation of `RingIO` in a separate file, because it really doesn't belong in the low-level ringbuf library code.

Can we put those helpers in `ringbuf.h` but named appropriately, eg `ringbuf_memcpy_get_internal`?  And a comment about why its "internal".

You can leave the `MP_ALWAYSINLINE` if it makes rp2 smaller.

andrewleech replied:
After the refactor I tested with/without `MP_ALWAYSINLINE` and get no change. Might also be realted to a different computer being used. Regardless, I don't think it's adding value anymore so have removed it.

andrewleech commented on docs/library/micropython.rst:
```diff
@@ -147,3 +147,66 @@ Functions
 
    There is a finite queue to hold the scheduled functions and `schedule()`
    will raise a `RuntimeError` if the queue is full.
+
```
@dpgeorge I've moved / updated the docs here... does this feel long enough now you'd prefer it in its own file?

dpgeorge replied:
I think it's OK to keep it in this file.

If more classes get added in the future then we can factor them out.

dpgeorge commented on py/objringio.h:
```diff
@@ -0,0 +1,37 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#ifndef MICROPY_INCLUDED_PY_OBJRINGIO_H
```
I don't think this header is necessary.  All types are declared in `py/obj.h`, so just put `mp_type_micropython_ringio` there.

This file is only needed if RingIO would export structs or helper functions.

dpgeorge commented on py/ringbuf.c:
```diff
@@ -83,17 +81,7 @@ int ringbuf_get_bytes(ringbuf_t *r, uint8_t *data, size_t data_len) {
-    }
-    memcpy(datap, r->buf + iget, iget_a - iget);
-    r->iget = iget_a;
+    ringbuf_memcpy_get_internal(r, data, data_len);
```
We are going to need to add some tests that cover this.  Probably in `ports/unix/coverage.c`.

andrewleech replied:
Thanks for the suggestion, I've added some extra tests to the ringbuf section there. 
Even though these aren't my functions ;-) I did modify them, making this a good opportunity to improve overall coverage.

dpgeorge replied:
Looks good now!

andrewleech commented:
Thanks @dpgeorge I've updated based on your most recent review. Lets see if the overall coverage is improved even more now!

dpgeorge commented on py/objringio.c:
```diff
@@ -0,0 +1,130 @@
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2019 Jim Mussared
```
Maybe this should be your copyright instead, if you authored this entire file yourself?

dpgeorge commented on docs/library/micropython.rst:
```diff
@@ -147,3 +147,66 @@ Functions
+   will allocate a 17 byte buffer internally so it can hold 16 bytes of data.
+   When passing in a pre-allocated buffer however one byte less than its
+   original length will be available for storage, eg. ``RingIO(bytearray(16))``
+   will only hold 15 bytes of data.
```
Maybe worth mentioning that reading and writing are thread and IRQ safe when used separately?

dpgeorge commented on ports/unix/coverage.c:
```diff
@@ -692,6 +692,24 @@ static mp_obj_t extra_coverage(void) {
+        mp_printf(&mp_plat_print, "%d\n", ringbuf_get_bytes(&ringbuf, get, 7));
+        mp_printf(&mp_plat_print, "%s\n", get);
+        // Prefill ringbuffer
+        for (int i = 0; i < 98; ++i) {
```
Instead of 98, maybe `sizeof(buf) - 2`?  Or even `sizeof(buf) - 3` to indicate that you allow 2 bytes free space, then the 7 added below will not fit.

dpgeorge commented on ports/unix/coverage.c:
```diff
@@ -692,6 +692,24 @@ static mp_obj_t extra_coverage(void) {
+        // Should fail - too full.
+        mp_printf(&mp_plat_print, "%d\n", ringbuf_put_bytes(&ringbuf, put, 7));
+        // Should fail - buffer too big.
+        uint8_t large[105] = {0};
```
Maybe `sizeof(buf) + 5`?

andrewleech commented:
They're helpful suggestions thanks @dpgeorge, all addressed in latest push.

dpgeorge commented on ports/unix/coverage.c:
```diff
@@ -692,6 +692,24 @@ static mp_obj_t extra_coverage(void) {
+        mp_printf(&mp_plat_print, "%d\n", ringbuf_get_bytes(&ringbuf, get, 7));
+        mp_printf(&mp_plat_print, "%s\n", get);
+        // Prefill ringbuffer
+        for (int i = 0; i < (sizeof(buf) - 3); ++i) {
```
int -> uint8_t, and no parenthesis around the sizeof-3 (for consistency)

andrewleech commented:
Thanks @dpgeorge that last build failure from the coverage addition has been fixed.

dpgeorge commented:
Thanks for updating, now merged!

AmirHmZz commented:
Thanks everyone for working on this!

ma261065 commented:
I got my first chance to use RingIO in a project the other day. I needed a 100kB buffer for some audio data and thought this would be perfect.

However, it seems that the size of the buffer is limited to 64kB by the use of uint16_t in this struct:

```
typedef struct _ringbuf_t {
    uint8_t *buf;
    uint16_t size;
    uint16_t iget;
    uint16_t iput;
} ringbuf_t;
```

So, I had to change them to uint32_t and compile a build, and then it worked just fine.

Is there a good reason that these were made unint16 and not uint32?

Could it be updated in the next release to allow larger buffers?

andrewleech commented:
The ringbuf C structure is used on some much smaller devices where every byte of ram counts, so it was originally made smaller as there were no large needs. 

Maybe a separate larger version of the struct could be defined and used instead for situations where a larger ringio object is created.

Actually that probably will overcomplicate things a bit much. Perhaps the size of the elements used in ringbuf struct could be defined by the ROM device size defines so on larger devices a uint32 is used. 

Regardless these changes will need a new PR, could you raise this as a new issue

ma261065 commented:
Thanks Andrew. I appreciate that we're generally running on resource-constrained devices, and so a good default would have been that "64kB should be enough for anybody".

Maybe I'm missing something, but as far as I understand the only change required is to change those three uint16_t to uint32_t, and this will work on small or large devices. Obviously if you try and allocate a ring buffer bigger than the memory available on the small device it will fail, but I don't see the need to have different versions for different devices.

Am I missing something?

(And, yes, I'll raise an issue as creating a PR is beyond my skills at this point)

andrewleech commented:
Changing these three elements from uint16 to unit32 is a global change in the firmware so will affect other existing ringbuf usages in C. Each ringbuf instance will then use an additional 6 bytes (2 per element in the struct). This might not seem like much but in the existing usages this is extra ram used for no feature gain.

ma261065 commented:
Fair enough - but isn't "The ability to create larger ring buffers" a feature gain?

And, unless I'm misunderstanding "affect other existing ringbuf usages" - this wouldn't be a breaking change would it?

Personally, I think it's worth the 6 extra bytes, but I'll defer the decision to you as the author.

dpgeorge commented:
I think the simplest thing to do would be make the `uint16_t` for the size configurable by the port.  And make it `uint32_t` only if the port has more than 64k of heap available (otherwise there's no point in it being larger).

ma261065 commented:
Issue raised - https://github.com/micropython/micropython/issues/16560
